File: kuramoto2d.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
# Morris-Lecar model on a 2D lattice
# FvW 03/2018

import os
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from scipy.interpolate import interp1d
import cv2

def kuramoto2d(N, T, t0, dt, s, K0, K1, sd_omega):
    s_sqrt_dt = s*np.sqrt(dt)
    omega = 1.0 + sd_omega * (np.random.rand(N,N) - 0.5)
    phi = 2*np.pi * np.random.rand(N,N) - np.pi
    zR = np.cos(phi)
    X = np.zeros((T,N,N))
    dphi = np.zeros((N,N))
    phi_up = np.zeros((N,N))
    phi_down = np.zeros((N,N))
    phi_left = np.zeros((N,N))
    phi_right = np.zeros((N,N))
    # coupling constant time course
    u = K0*np.ones(t0)
    v = np.linspace(K0,K1,num=T)
    K = np.hstack((u,v)) # 1-dim time course of K
    K4 = K/4
    # iterate
    for t in range(t0+T):
        if (t%100 == 0): print(f"    t = {t:d}/{t0+T:d}\r", end="")
        phi_up = np.roll(phi, -1, axis=0)
        phi_down = np.roll(phi, 1, axis=0)
        phi_left = np.roll(phi, -1, axis=1)
        phi_right = np.roll(phi, 1, axis=1)
        dphi = np.zeros((N,N))
        dphi += (np.sin(phi_up-phi) + np.sin(phi_down-phi))
        dphi += (np.sin(phi_left-phi) + np.sin(phi_right-phi))
        dphi = omega + K4[t]*dphi
        # stochastic integration
        phi += (dphi*dt + s_sqrt_dt*np.random.randn(N,N))
        if (t >= t0):
            #X[t-t0,:,:] = np.cos(phi)
            X[t-t0,:,:] = np.angle(np.exp(1j*phi)) # phase
    print("\n")
    return X


def animate_pyplot1(fname, data, downsample=10):
    """
    Animate 3D array as .mp4 using matplotlib.animation.FuncAnimation
    modified from:
    https://matplotlib.org/stable/gallery/animation/simple_anim.html
    (Faster than animate_pyplot2)

    Args:
        fname : string
            save result in mp4 format as `fname` in the current directory
        data : array (nt,nx,ny) (time, space, space)
        downsample : integer, optional
            temporal downsampling factor

    """
    nt, nx, ny = data.shape
    n1 = nt
    if downsample:
        n1 = int(nt/downsample) # number of samples after downsampling
        t0 = np.arange(nt)
        f_ip = interp1d(t0, data, axis=0, kind='linear')
        t1 = np.linspace(0, nt-1, n1)
        data = f_ip(t1)
    vmin, vmax = data.min(), data.max()
    # setup animation image
    cm = plt.cm.gray # plt.cm.jet # plt.cm.gray
    fig = plt.figure(figsize=(6,6))
    ax = plt.gca()
    ax.axis("off")
    t = plt.imshow(data[0,:,:], origin="lower", cmap=cm,
                   interpolation="bilinear", vmin=vmin, vmax=vmax)
    plt.tight_layout()
    # frame generator
    print("[+] Animate")
    def animate(i):
        if (i%10 == 0): print(f"    i = {i:d}/{n1:d}\r", end="")
        t.set_data(data[i,:,:])
    # create animation
    ani = animation.FuncAnimation(fig, animate, frames=n1, interval=10)
    #ani.save(fname)
    writer = animation.FFMpegWriter(fps=15, bitrate=1200)
    ani.save(fname, writer=writer)
    plt.show()
    print("\n")


def animate_pyplot2(fname, data, downsample=10):
    """
    Animate 3D array as .mp4 using matplotlib.animation.ArtistAnimation
    modified from:
    https://matplotlib.org/stable/gallery/animation/dynamic_image.html
    (Slower than animate_pyplot1)

    Args:
        fname : string
            save result in mp4 format as `fname` in the current directory
        data : array (nt,nx,ny) (time, space, space)
        downsample : integer, optional
            temporal downsampling factor

    """
    nt, nx, ny = data.shape
    n1 = nt
    if downsample:
        n1 = int(nt/downsample) # number of samples after downsampling
        t0 = np.arange(nt)
        f_ip = interp1d(t0, data, axis=0, kind='linear')
        t1 = np.linspace(0, nt-1, n1)
        data = f_ip(t1)
    print("[+] Animate")
    vmin, vmax = data.min(), data.max()
    fig = plt.figure(figsize=(6,6))
    ax = plt.gca()
    ax.axis("off")
    plt.tight_layout()
    ims = []
    for i in range(n1):
        if (i%10 == 0): print(f"    i = {i:d}/{n1:d}\r", end="")
        im = ax.imshow(data[i,:,:], origin="lower", cmap=plt.cm.gray, \
                       vmin=vmin, vmax=vmax, animated=True)
        if i == 0:
            ax.imshow(data[i,:,:], origin="lower", cmap=plt.cm.gray, \
                           vmin=vmin, vmax=vmax)
        ims.append([im])
    ani = animation.ArtistAnimation(fig, ims, interval=50, blit=True)
    #ani.save(fname2)
    writer = animation.FFMpegWriter(fps=15, bitrate=1200)
    ani.save(fname, writer=writer)
    plt.show()
    print("\n")


def animate_video(fname, x, downsample=10):
    nt, nx, ny = x.shape
    n1 = nt
    if downsample:
        n1 = int(nt/downsample) # number of samples after downsampling
        t0 = np.arange(nt)
        f_ip = interp1d(t0, x, axis=0, kind='linear')
        t1 = np.linspace(0, nt-1, n1)
        x = f_ip(t1)
    # BW
    y = 255 * (x-x.min()) / (x.max()-x.min())
    # BW inverted
    #y = 255 * ( 1 - (x-x.min()) / (x.max()-x.min()) )
    y = y.astype(np.uint8)
    #print(f"n1 = {n1:d}, nx = {nx:d}, ny = {ny:d}")
    # write video using opencv
    frate = 30
    print("[+] Animate")
    out = cv2.VideoWriter(fname, \
                          cv2.VideoWriter_fourcc(*'mp4v'), \
                          frate, (nx,ny))
    for i in range(n1):
        print(f"    i = {i:d}/{n1:d}\r", end="")
        img = np.ones((nx, ny, 3), dtype=np.uint8)
        for j in range(3): img[:,:,j] = y[i,::-1,:]
        out.write(img)
    out.release()
    print("")


def main():
    print("Kuramoto lattice model\n")
    N = 100 # 128
    T = 2500
    t0 = 500
    dt = 0.05
    s = 0.1
    # Kuramoto parameters
    K0 = 1.0
    K1 = 5.0
    sd_omega = 0.25
    print(f"[+] Lattice size N: {N:d}")
    print(f"[+] Time steps T: {T:d}")
    print(f"[+] Warm-up steps t0: {t0:d}")
    print(f"[+] Integration time step dt: {dt:.4f}")
    print(f"[+] Noise intensity: {s:.2f}")
    print(f"[+] Coupling constant: K0={K0:.2f} ... K1={K1:.2f}")
    print(f"[+] Frequency distribution std. dev.: {sd_omega:.3f}")

    # run simulation
    data = kuramoto2d(N, T, t0, dt, s, K0, K1, sd_omega)
    #print("[+] Data dimensions: ", data.shape)

    ''' plot mean signal
    m = np.mean(np.reshape(data, (T,N*N)), axis=1)
    plt.figure(figsize=(12,4))
    plt.plot(m, "-k")
    plt.tight_layout()
    plt.show()
    '''

    # save data
    #fname1 = f"kuramoto2d_K_{K0:.2f}_{K1:.2f}_s_{s:.2f}.npy"
    #np.save(fname1, data)
    #print(f"[+] Data saved as: {fname1:s}")

    # video
    fname2 = f"kuramoto2d_K_{K0:.2f}_{K1:.2f}_s_{s:.2f}.mp4"
    #animate_pyplot1(fname2, data, downsample=5)
    #animate_pyplot2(fname2, data, downsample=5)
    animate_video(fname2, data, downsample=5) # fastest
    print(f"[+] Animation saved as: {fname2:s}")


if __name__ == "__main__":
    os.system("clear")
    main()

File: kuramoto2d.jl
#!/usr/local/bin/julia
# last tested Julia version: 1.6.1
# Kuramoto model on a 2D lattice
# FvW 03/2018

using Interpolations
using NPZ
using PyCall
using PyPlot
using Statistics
using VideoIO
@pyimport matplotlib.animation as anim

function kuramoto2d(N, T, t0, dt, s, K0, K1, sd_omega)
    s_sqrt_dt = s*sqrt(dt)
	omega = 1.0 .+ sd_omega * (rand(N,N) .- 0.5)
    phi = 2pi .* rand(N,N) .- pi
	X = zeros(Float64,T,N,N)
	dphi = zeros(Float64,N,N)
	phi_up = zeros(Float64,N,N)
	phi_down = zeros(Float64,N,N)
	phi_left = zeros(Float64,N,N)
    phi_right = zeros(Float64,N,N)
    # coupling constant time course
    u = K0*ones(t0)
    v = [_ for _ in range(K0,stop=K1,length=T)]
    K = vcat(u,v) # 1-dim time course of K
    K4 = K .* 0.25
    # iterate
    for t in range(1, stop=t0+T, step=1)
        (t%100 == 0) && print("    t = ", t, "/", t0+T, "\r")
		phi_up = circshift(phi, [-1 0])
        phi_down = circshift(phi, [1 0])
        phi_left = circshift(phi, [0 -1])
        phi_right = circshift(phi, [0 1])
        dphi = zeros(Float64,N,N)
	    dphi += (sin.(phi_up-phi) + sin.(phi_down-phi))
	    dphi += (sin.(phi_left-phi) + sin.(phi_right-phi))
        dphi = omega .+ K4[t] .* dphi
		# stochastic integration
		phi += (dphi*dt + s_sqrt_dt*randn(N,N))
		#(t > t0) && (X[t-t0,:,:] = cos.(phi))
        (t > t0) && (X[t-t0,:,:] = angle.(exp.(im.*phi))) # angle
    end
    println("\n")
    return X
end

function animate_pyplot(fname, data)
    """
    Animate 3D array as .mp4 using PyPlot, save as `fname`
    array dimensions:
        1: time
        2, 3: space
    """
    nt, nx, ny = size(data)
    vmin = minimum(data)
    vmax = maximum(data)
    # setup animation image
    fig = figure(figsize=(6,6))
    axis("off")
    t = imshow(data[1,:,:], origin="lower", cmap=ColorMap("gray"),
			   vmin=vmin, vmax=vmax)
    tight_layout()
    # frame generator
    println("[+] animate")
    function animate(i)
        (i%100 == 0) && print("    t = ", i, "/", nt, "\r")
        t.set_data(data[i+1,:,:])
    end
    # create animation
    ani = anim.FuncAnimation(fig, animate, frames=nt, interval=10)
    println("\n")
    # save animation
    ani[:save](fname, bitrate=-1,
               extra_args=["-vcodec", "libx264", "-pix_fmt", "yuv420p"])
    show()
end

function animate_video(fname, data, downsample=10)
    """
    Animate 3D array as .mp4 using VideoIO, save as `fname`
    array dimensions:
        1: time
        2, 3: space
    """
    nt, nx, ny = size(data)
    n1 = nt
    if (downsample > 0)
        println("[+] Downsampling ", downsample)
        n1 = Int(nt/downsample) # number of samples after downsampling
        data1 = zeros(Complex64,n1,nx,ny)
        t0 = range(1,stop=nt)
        t1 = range(1,stop=nt,length=n1)
        for i in range(1,stop=nx)
            for j in range(1,stop=ny)
                y = data[:,i,j]
                f_ip = LinearInterpolation(t0, y)
                data1[:,i,j] = f_ip(t1)
            end
        end
    else
        data1 = copy(data)
    end
    # BW
    y = UInt8.(round.(255*(data1 .- minimum(data1)) /
                           (maximum(data1)-minimum(data1))))
    # BW inverted
    #y = UInt8.(round.(255 .- 255*(data1 .- minimum(data1)) /
    #                              (maximum(data1)-minimum(data1))))
    encoder_options = (color_range=2, crf=0, preset="medium")
    framerate=30
    T = size(data1,1)
    println("[+] Animate:")
    open_video_out(fname, y[1,end:-1:1,:], framerate=framerate,
                   encoder_options=encoder_options) do writer
        for i in range(2,stop=n1,step=1)
            (i%100 == 0) && print("    i = ", i, "/", T, "\r")
            write(writer, y[i,end:-1:1,:])
        end
    end
    println("\n")
end

function main()
    println("Kuramoto lattice model\n")
    N = 128
    T = 3000
    t0 = 0
    dt = 0.05
    s = 0.2
    # Kuramoto parameters
    K0 = 0.5
    K1 = 4.0
    sd_omega = 0.25
    println("[+] Lattice size N: ", N)
    println("[+] Time steps T: ", T)
    println("[+] Warm-up steps t0: ", t0)
    println("[+] Integration time step dt: ", dt)
    println("[+] Noise intensity s: ", s)
	println("[+] Coupling constant K0...K1: ", K0, " ... ", K1)
    println("[+] Frequency distribution std. dev.: ", sd_omega)

    # run simulation
    data = kuramoto2d(N, T, t0, dt, s, K0, K1, sd_omega)
    println("[+] Data dimensions: ", size(data))

    # plot mean signal
    m = mean(reshape(data, (T,N*N)), dims=2)
    plot(m, "-k"); show()

    # save data
    K0_str = rpad(K0, 4, '0') # coupl. param. 4-char string
    K1_str = rpad(K1, 4, '0') # coupl. param. 4-char string
    s_str = rpad(s, 4, '0') # noise as 4-char string
    fname1 = string("kuramoto2d_K_", K0_str, "_", K1_str, "_s_", s_str, ".npy")
    #npzwrite(fname1, data)
    #println("[+] Data saved as: ", fname1)

	# video
    fname2 = string("kuramoto2d_K_", K0_str, "_", K1_str, "_s_", s_str, ".mp4")
    #animate_pyplot(fname2, data) # slow
    animate_video(fname2, data, 5) # fast
    println("[+] Data saved as: ", fname2)
end

main()
